client/request.go

```diff
@@ -104,12 +104,8 @@ func (cli *Client) buildRequest(method, path string, body io.Reader, headers htt
 	req = cli.addHeaders(req, headers)
 
 	if cli.proto == "unix" || cli.proto == "npipe" {
-		// For local communications, it doesn't matter what the host is. We just
-		// need a valid and meaningful host name. For details, see:
-		//
-		// - https://github.com/docker/engine-api/issues/189
-		// - https://github.com/golang/go/issues/13624
-		req.Host = "docker"
+		// For local communications, it doesn't matter what the host is.
+		req.Host = DummyHost
 	}
 
 	req.URL.Host = cli.addr
```

Comment:
```suggestion
		// For local communications, the host is undefined.
		req.Host = ""
		req.URL.Host = ""
	} else {
		req.URL.Host = cli.addr
	}
```
[Go's HTTP client is perfectly happy to build a request with an empty Host header.](https://go.dev/play/p/9nIgcpBT96q) The value of the header defaults to `req.URL.Host` if `req.Host` is the empty string, so both need to be zeroed out.

Reply:
Ah, I was playing with the same idea. Honestly, it's such a mess that I didn't  try, because I was afraid something could depend on it being non-empty.

Let me give it a try once we have the BuildKit fixes in, then we can get a fresh run of CI

Reply:
Sadly, [`http.Transport` insists that `req.URL.Host != ""`.](https://github.com/golang/go/blob/6244b1946bc2101b01955468f1be502dbadd6807/src/net/http/transport.go#L558-L569) It also insists that `req.URL.Scheme == "http" || req.URL.Scheme == "https"` so it makes sense that it is so opinionated: it is the transport for those schemes and those schemes only, so it enforces the semantics of HTTP(S) over TCP. It looks like we would have to write a custom `http.RoundTripper` for http+unix and http+npipe to get the "empty host is okay" semantics. The path of least resistance sadly is looking to be to use some arbitrary non-empty string as the host.

Reply:
Hmm


But it also means that we can potentially check for http(s) schemes instead of "unix" or "npipe" ? 🤔